src/main/java/org/jnode/fs/xfs/directory/BPlusTreeDirectory.java

```diff
@@ -121,4 +99,35 @@ public BPlusTreeDirectory(byte [] data, long offset, long iNodeNumber, XfsFileSy
         final NodeDirectory leafDirectory = new NodeDirectory(buffer.array(), 0, fileSystem, iNodeNumber, extents, leafExtentIndex);
         return leafDirectory.getEntries(parentDirectory);
     }
+
+
+    public List<DataExtent> getFlattenedExtents(byte[] data,int level,List<DataExtent> currentExtents, int btreeBlockOffset, int numrecs) throws IOException {
+        boolean isBtreeExtentList = false;
+        if (btreeBlockOffset == 0){
+            numrecs = BigEndian.getUInt16(data,6);
+            final long signature = BigEndian.getUInt32(data, 0);
+            isBtreeExtentList = signature == 0x424D4133 || signature == 0x424D4150;
+            final int filesystemOffset = fileSystem.getXfsVersion() == 5 ? 64 : 24;
+            btreeBlockOffset = (int) (fileSystem.getSuperblock().getBlockSize() + filesystemOffset) / 2;
+        }
+        for (int i = 0; i < numrecs; i++) {
+            final long fsBlockNo = isBtreeExtentList ? BigEndian.getInt64(data, btreeBlockOffset) : BigEndian.getUInt32(data, btreeBlockOffset);
```

Comment:
I'd recommend not using `final` for local variables, in preparation for code style in Engine code.